crates/autopilot/src/decoded_settlement.rs

```diff
@@ -0,0 +1,511 @@
+//! This module contains the logic for decoding the function input for
+//! GPv2Settlement::settle function.
+
+use {
+    anyhow::{anyhow, Context, Result},
+    bigdecimal::{Signed, Zero},
+    contracts::GPv2Settlement,
+    ethcontract::{common::FunctionExt, tokens::Tokenize, Address, Bytes, U256},
+    model::order::{Order, OrderKind},
+    num::BigRational,
+    number_conversions::big_rational_to_u256,
+    shared::{conversions::U256Ext, external_prices::ExternalPrices},
+    web3::ethabi::{Function, Token},
+};
+
+// Original type for input of `GPv2Settlement.settle` function.
+type DecodedSettlementTokenized = (
+    Vec<Address>,
+    Vec<U256>,
+    Vec<(
+        U256,            // sellTokenIndex
+        U256,            // buyTokenIndex
+        Address,         // receiver
+        U256,            // sellAmount
+        U256,            // buyAmount
+        u32,             // validTo
+        Bytes<[u8; 32]>, // appData
+        U256,            // feeAmount
+        U256,            // flags
+        U256,            // executedAmount
+        Bytes<Vec<u8>>,  // signature
+    )>,
+    [Vec<(Address, U256, Bytes<Vec<u8>>)>; 3],
+);
+
+#[derive(Debug)]
+#[allow(dead_code)]
+pub struct DecodedSettlement {
+    // TODO check if `EncodedSettlement` can be reused
```

Comment:
I think I could reuse `EncodedSettlement` for this, but did not want to move out EncodedSettlement out of `solver` crate and make even bigger refactor change. But if you agree, I can try to merge these two structs at a later stage.